find_exponential/src/main.rs

```diff
@@ -0,0 +1,197 @@
+        .filter(|e| e.file_type().is_file())
+        .filter(|e| if let Some(ext) = e.path().extension() { ext == "rs" } else { false });
+
+    let mut literals = HashSet::new();
+    for file in walkdir {
+        if file.path() == Path::new("../src/entities.rs") {
+            continue;
+        }
+        println!("parsing {}", file.path().display());
+        let content = fs::read_to_string(file.path())
+            .expect(&format!("unable to read file {:?}", file.path()));
+        let parsed = syn::parse_file(&content)
+            .expect(&format!("unable to parse file {:?}", file.path()));
+        literals::extract_literals_from_items(&mut literals, parsed.items);
+    }
+
+    let literals: Vec<_> = literals.into_iter()
+        .filter(|lit| !lit.contains(&b'\n'))
+        .filter(|lit| !lit.is_empty())
+        .collect();
+
+    let pool = ThreadPool::new(num_cpus);
+    let mut count = -(num_cpus as isize) * BATCH_SIZE as isize;
+    let start_time = Instant::now();
+    let (tx, rx) = mpsc::channel();
+    for _ in 0..num_cpus {
+        tx.send(()).unwrap();
+    }
+    loop {
+        rx.recv().unwrap();
+        let mut literals = literals.clone();
+        let tx = tx.clone();
+        count += BATCH_SIZE as isize;
+        pool.execute(move || {
+            literals.shuffle(&mut rand::thread_rng());
+            let combs = literals.iter()
+                .combinations(COMBINATIONS)
+                .take(BATCH_SIZE);
+            for comb in combs {
+                let res = panic::catch_unwind(|| {
+                    let concatenated = comb.into_iter().flatten().cloned().collect::<Vec<_>>();
+                    let s = String::from_utf8(concatenated).unwrap();
+                    test(&s);
+                });
+                if res.is_err() {
+                    ::std::process::exit(1);
+                }
+            }
+            tx.send(()).unwrap();
+        });
+        if count > 0 {
+            println!("{:<20} throughput: {}", count, count as u64 / start_time.elapsed().as_secs());
+        }
+    }
+}
+
+fn test(pattern: &str) {
+    let (_, _, non_linear) = test_pattern(pattern, SAMPLE_SIZE, false);
+    if non_linear {
+        let (array, score, non_linear) = test_pattern(pattern, SAMPLE_SIZE*2, true);
+        if non_linear {
+            print_find(pattern, &array, score);
+        }
+    }
+}
+
+fn test_pattern(pattern: &str, sample_size: usize, remove_outliers: bool) -> (Array2<f64>, f64, bool) {
+    let s = pattern.repeat(NUM_BYTES / pattern.len());
+    let mut array = Array2::zeros((2, sample_size));
+
+    let mut i = 0;
+    while i < sample_size {
+        let (dur, n) = calculate_point(&s, i+1, sample_size);
+        array[[0, i]] = n as f64;
+        array[[1, i]] = dur.as_nanos() as f64;
+        if DEBUG_LEVEL >= 3 {
+            println!("duration: {}", dur.as_nanos());
+        }
+
+        if remove_outliers && i > 0 && array[[1, i-1]] > array[[1, i]] {
```

Comment:
I try to keep the throughput as high as possible. As such, I first do a fast test, to filter out most non-linear patterns. However rescheduling / spikes result in skewed values. As such if the first test detects possible non-linear behaviour, I reperform the test, but with higher precision. The second test is the first test, but with double as many samples. Additionally, if I encounter an outlier, which is any value that is larger than its successor (because I expect a strict growth of parsing time with increasing length), I recompute that value. While this is rather expensive, it highly reduces false positives due to rescheduling.
Unfortunately it's not always true that parsing time increases strictly, as seen in #287. Therefore, I have now implemented detection of large outliers (at least 50× as large as their successor), which I try to recompute ten times. If they are a large outlier all those ten times, I mark that pattern as weird pattern (non necessarily non-linear, but at least it has weird parsing behaviour).

>  Ideally, we'd use a method that is resistant to spurious outliers.

Outliers aren't uncommon unfortunately due to rescheduling. If you have an idea how to process outliers in the scoring function, feel free to tell me and I'll implement it :) As it is right now it is able to handle some outliers pretty well (the first test can contain a few outliers as long as they aren't too drastic). If it fails however, I'll manually remove the outliers in the second test.

I'll add a comment explaining this.